malariagen_data/anoph/hap_freq.py

```diff
@@ -0,0 +1,393 @@
+        self._cache_annotator = None
+
+    @check_types
+    @doc(
+        summary="""
+            Compute haplotype frequencies for a region.
+        """,
+        returns="""
+            A dataframe of haplotype frequencies, one row per haplotype.
+        """,
+        notes="""
+            Cohorts with fewer samples than `min_cohort_size` will be excluded from
+            output data frame.
+        """,
+    )
+    def haplotypes_frequencies(
+        self,
+        region: base_params.region,
+        cohorts: base_params.cohorts,
+        sample_query: Optional[base_params.sample_query] = None,
+        min_cohort_size: base_params.min_cohort_size = 10,
+        sample_sets: Optional[base_params.sample_sets] = None,
+        chunks: base_params.chunks = base_params.native_chunks,
+        inline_array: base_params.inline_array = base_params.inline_array_default,
+    ) -> pd.DataFrame:
+        # Access sample metadata.
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets, sample_query=sample_query
+        )
+
+        # Build cohort dictionary, maps cohort labels to boolean indexers.
+        coh_dict = locate_cohorts(cohorts=cohorts, data=df_samples)
+
+        # Remove cohorts below minimum cohort size.
+        coh_dict = {
+            coh: loc_coh
+            for coh, loc_coh in coh_dict.items()
+            if np.count_nonzero(loc_coh) >= min_cohort_size
+        }
+
+        # Early check for no cohorts.
+        if len(coh_dict) == 0:
+            raise ValueError(
+                "No cohorts available for the given sample selection parameters and minimum cohort size."
+            )
+
+        # Access haplotypes.
+        ds_hap = self.haplotypes(
+            region=region,
+            sample_sets=sample_sets,
+            sample_query=sample_query,
+            chunks=chunks,
+            inline_array=inline_array,
+        )
+
+        # Early check for no SNPs.
+        if ds_hap.sizes["variants"] == 0:  # pragma: no cover
+            raise ValueError("No SNPs available for the given region.")
+
+        # Access genotypes.
+        gt = allel.GenotypeDaskArray(ds_hap["call_genotype"].data)
+        with self._dask_progress(desc="Compute haplotypes"):
+            gt = gt.compute()
+
+        # Count haplotypes.
+        freq_cols = dict()
+        cohorts_iterator = self._progress(
+            coh_dict.items(), desc="Compute allele frequencies"
+        )
+        hap_track: dict[np.int64, float] = {}
+        for coh, loc_coh in cohorts_iterator:
+            hap_track = {k: 0 for k in hap_track.keys()}
+            n_samples = np.count_nonzero(loc_coh)
+            assert n_samples >= min_cohort_size
+            gt_coh = allel.GenotypeDaskArray(da.compress(loc_coh, gt, axis=1))
+            gt_hap = gt_coh.to_haplotypes().compute()
+            f, _, _ = haplotype_frequencies(gt_hap)
+            hap_track.update(f)
+            freq_cols["frq_" + coh] = list(hap_track.values())
```

Comment:
I am a little concerned about ensuring that the frequency values are the same for each cohort, i.e., the values correspond to the same haplotypes. Also I think this could end up bleeding values between different cohorts.  

Am I right that you are trying to use `hap_track` to keep a record of what haplotypes you have seen so far?

Consider what would happen here if you have two cohorts, where some haplotypes are present in both cohorts, and each cohort has haplotypes which are not observed in the other cohort.

I wonder if it would be safer to do this in two passes. In the first pass, you compute the results of calling `haplotype_frequencies()` for each cohort. Then you find the set of all haplotype keys across all cohorts, and use this to generate a consistent array of frequencies for each cohort, guaranteeing the same order of values for each cohort, and filling with 0 in case the haplotype is not observed.

Reply:
If you did that you could also add a "label" column to the output dataframe, which could contain the hash for each haplotype. Not super user friendly, but would at least allow us to refer to the different haplotypes in some way?

Reply:
Not sure I get what you mean. I have little time so what I say might be gibberish. 

For the first part, `haplotype_frequencies()` returns a dict where the keys are the hash for each haplotype and the value is the frequency. These keys are then used by `hap_track` so it should already do what I think you asked with one pass.  
For the second part, "label" is already used for the index (because `plot_frequencies_heatmap()` requires it, I think). We could keep the old index (containing the hash) in a column "hash" if you want.

Reply:
> For the first part, `haplotype_frequencies()` returns a dict where the keys are the hash for each haplotype and the value is the frequency. These keys are then used by `hap_track` so it should already do what I think you asked with one pass.

Right, after you have iterated through all cohorts then `hap_track` will have seen all haplotypes present in any cohort, and so the keys will contain hashes of all haplotypes present.

But if you are outputting the frequencies as you go, within that first pass, then the number of values you find will vary between cohorts. This is because you find more haplotypes as you visit more cohorts.

Also, `hap_track` currently accumulates data from all cohorts. When you visit a new cohort, any haplotypes previously found will have their frequencies updated. But any haplotypes found in previous cohorts but not present in the current cohort will still be present within `hap_track`. If you then take all values from `hap_track` you could get a mangle of data from the current cohort and data from cohorts visited in previous iterations.

Reply:
Sorry @alimanfoo, I thought I had commented further yesterday but my comments seems to have got lost (I probably forgot to click on the big green button). 

> Right, after you have iterated through all cohorts then hap_track will have seen all haplotypes present in any cohort, and so the keys will contain hashes of all haplotypes present.
But if you are outputting the frequencies as you go, within that first pass, then the number of values you find will vary between cohorts. This is because you find more haplotypes as you visit more cohorts.

That's right. This is the reason why there is a step before creating the DataFrame that adds 0 to the end of the lists that are shorter than the total number of haplotypes (which is the length of the last list generated). This could be a problem if `update` doesn't treat the dict entries as a queue and reorders them (according to whatever order it likes). I tried this:
```
d = {"B": 0, "C":1}
d.update({"A": 2, "B":3}
d.values()
```
and it correctly returns `[3,1,2]` so the order seems to be indeed kept. So the new haplotypes are always added at the end of `hap_track`. I should probably document my code better (that is, at least a little) to make my reasoning clearer.

> Also, hap_track currently accumulates data from all cohorts. When you visit a new cohort, any haplotypes previously found will have their frequencies updated. But any haplotypes found in previous cohorts but not present in the current cohort will still be present within hap_track. If you then take all values from hap_track you could get a mangle of data from the current cohort and data from cohorts visited in previous iterations.

The first step of every cohort run is to reset all values of `hap_track` to 0 so I don't think there is a risk of mangling. I was scared at some point that the content of `hap_track` would be copied by reference instead of by value in `freq_coh` but given that they end up different, I don't think that's the case. 

There is one situation that I am not quite sure how to handle though. Resetting all values to 0 for each cohort run makes sense for `freq` and `count` but, if a haplotype is not observed, having 0 as the `nobs` value might be wrong. What do you think? Should the lengthening step set all values in `nobs` to be equal or is 0 for haplotypes that are not observed OK. As far as I can see, it only matters for the CIs ... and I just convinced myself that it is wrong to have `nobs == 0` so I will update that when I come back from my leave.